### ISSUE TITLE
partsToPosix should return missing or ambiguous times appropriately

### ISSUE THREAD

**Issue by malaire** [justinmimbs/time-extra#3]
The `partsToPosix` function doesn't work correctly because it always returns a single `Time.Posix`.

When near DST change, time conversion like this can result in zero or two `Time.Posix` values instead of one.

**Comment by justinmimbs** [comment]
Hi Markus,

Indeed. This decision to always return one `Posix` value is a practical choice, and the tradeoff is that the function tacitly handles bad input.

- When given parts describing an ambiguous or missing time (near a DST switch), the function returns the approximation that's closest to UTC time.
- When given out-of-range parts, the function clamps them to valid values.

It seems reasonable to have a separate, strict version of this function, returning zero, one, or two `Posix` values, which would allow the caller to handle bad input. Do you currently have a use case requiring such a function?

Thanks,

Justin

**Comment by malaire** [comment]
I don't currently have immediate use case, but I'm thinking about an application which would allow arbitrary conversions between timezones.

"When given parts describing an ambiguous or missing time (near a DST switch), the function returns the approximation that's closest to UTC time."

What does that mean in case where there are two possible UTC times? I'm thinking that `partsToPosix` should at least clearly document what it returns, e.g. always returning earliest possible UTC time if there are two possibilities.

**Comment by justinmimbs** [comment]
Here's what I mean by "closest to UTC time". If the time is ambiguous because it occurred twice during a fall back, then one of them will have less offset from UTC.

For example, if local time is ahead of UTC, then this is the later of the two:

```
                  UTC  | local
                  ---- | -----------
                  0:00 | 2:00 UTC+2
offset change ->  1:00 | 2:00 UTC+1  <- less offset
```

And if local time is behind UTC, then this is the earlier of the two:

```
                  UTC  | local
                  ---- | -----------
                  3:00 | 2:00 UTC-1  <- less offset
offset change ->  4:00 | 2:00 UTC-2
```

(If you think of "taking the earlier time" as like flooring a number, then you can think of "taking the closer to UTC" as like rounding toward zero.)

I agree that the documentation should explain this behavior (as it explains the clamping of parts), so I'll include it in the next update.

Thanks,

Justin

**Comment by malaire** [comment]
ok, as for missing times, I think it would actually be best as a separate function, so it doesn't complicate the common use case.

Something like `isValid : Zone -> Parts -> ???` which would check for e.g. that hour is not over 24, date is not February 29th unless on leap year, and full date/time is not during missing time at DST change.

I'm not sure whether that should return just `True`/`False`, or a type explaining exact reason why time is not valid.

**Comment by justinmimbs** [comment]
I think having a separate, strict version of `partsToPosix` would subsume the need for an `isValid` function. A strict version of `partsToPosix` could return a `Result Problem Posix`, where `Problem` is something like

```elm
type Problem
    = OutOfRange
    | MissingTime
    | AmbiguousTime ( Posix, Offset ) ( Posix, Offset )
```

and from that you could decide if the parts are valid for your needs or not. What do you think?

**Comment by malaire** [comment]
Yes, that sounds better. Although I'd prefer something like the following instead of using `Result` as I don't think `AmbiguousTime` is really an error here.

Also if `AmbiguousTime` is returning `Offset`, `UnambiguousTime` could do the same.

    type ComplexTime
        = OutOfRange
        | MissingTime
        | UnambiguousTime ( Posix, Offset )
        | AmbiguousTime ( Posix, Offset ) ( Posix, Offset )

**Comment by justinmimbs** [comment]
On second thought, I think `AmbiguousTime` can just provide the two `Posix` times, because the user can still get both offsets with `toOffset` if they want.

You're saying that including `AmbiguousTime` as part of the `Err` case unfairly calls it an "error," since it might not be an error as far as the user is concerned? That makes sense. I do like the clarity of the `Result` type, but I can see benefits of both ways.

**Comment by DavidDTA** [comment]
I have a use case for this (a time tracking application).

I think the API I would want is:
```elm
partsToPosixSafe : Zone -> Parts -> List { posix: Posix, offset: Offset }
```

My use case doesn't care too much about invalid dates, so I just grouped them all together (as the empty list) because I wasn't sure what the meanings of `OutOfRange` and `MissingTime` are. We could modify the signature if the distinction is important to others.

I went with a list because timezones can do wacky things. There don't appear to be any cases in the IANA database where there are 3 ambiguous datestimes for a single posix, but the definition of a custom timezone in Elm does not rule it out. With this signature, it is still easy to pattern match on the one-time and two-time cases.
